### ISSUE TITLE
Bad error message in pattern matching where variable name collides with constructor name

### ISSUE THREAD

**Issue by cpitclaudel** [coq/coq#11087]
#### Description of the problem

The following produces a very puzzling error message:

```coq
Inductive variable := index | datapoint (index: nat).

Fail Definition initial_value (v: variable)
  : match v with
    | index => nat
    | datapoint _ => nat
    end :=
  match v with
  | index => 0 (* Error here *)
  | datapoint index => 0
  end.

(*
In environment
v : variable
The term "0" has type "nat" while it is expected to have type
 "match ?v@{v1:=index} with
  | index | _ => nat
  end".
*)
```

Contrast with this, where the error message is much better:

```coq
Fail Definition initial_value (v: variable) : nat :=
  match v with
  | index => 0
  | datapoint index (* Error here *) => 0
  end.

(*
Found a constructor of inductive type variable while a constructor of nat is expected.
*)
```

The problem stems from forgetting that `index` is a constructor and using it as a binder name.  

#### Coq Version

8.9.1

**Comment by cpitclaudel** [comment]
In the same vein, this is puzzling:

```
Fail Definition initial_value (v: test)
  : match v with
    | a => nat
    | b _ => nat
    end :=
  match v with
  | idx => 0 (* Error here *)
  | b _ => false
  end.

(* Error: *)
(* In environment *)
(* v : test *)
(* The term "0" has type "nat" while it is expected to have type *)
(*  "match ?t@{v:=v; v1:=v} with *)
(*   | a | _ => nat *)
(*   end". *)
```

Shouldn't the error be on the `false`, rather than the 0?

**Comment by silene** [comment]
> Shouldn't the error be on the false, rather than the 0?

Coq stops as soon as it detects a typing error, so it does not even reach the point where it encounters `false`. My guess is that the `return` clauses is incorrectly elaborated in case of name collision.

**Comment by cpitclaudel** [comment]
> Coq stops as soon as it detects a typing error 

But which error? That 0 is valid, right?

**Comment by silene** [comment]
As Coq rightfully complains, the result `0` of the first branch does not have type `match ?v@{v1:=index} with index | _ => nat end`. If you write an explicit `return` clause in your match, then Coq no longer infers this broken clause, thus the first branch correctly typechecks, and once it reaches the second branch, Coq complains that it has `found a constructor of inductive type variable while a constructor of nat is expected`.

To summarize, the error message is fine. It is just the symptom of an earlier bug.

**Comment by cpitclaudel** [comment]
I think we're mixing the two errors.  When I said the error should be on false rather than on 0, I meant on the second example.

In the first one, do you mean an earlier bug in clause inference, or an earlier bug in the code example itself?

**Comment by silene** [comment]
Both your examples behave in the exact same way. The value of the first branch does not match its inferred type, so Coq complains about a type error. It never even reaches the second branch, where it would complain about `ident` (first example) or `false` (second example).

**Comment by cpitclaudel** [comment]
Why does not Coq use the annotation that's on the definition?  It seems to do that when the annotation is simpler (just nat instead of a match, like in the second example).

**Comment by JasonGross** [comment]
As I understand it, the issue here is that Coq checks the typing of each branch before checking that the patterns of subsequent branches are valid.  I believe @cpitclaudel is claiming that the error messages would be more clear if Coq first checked that there were no redundant branches and that all of the constructors in the branch patterns had the right types, and only after that checked that the branch cases themselves were well-typed and tried to unify with the return clause.  I agree.  (I will note that this is not so easy, though, because the elaboration of patterns in deep pattern matches can depend on typing constraints in the branches.  So I think what should actually happen is that Coq should gather all the error messages from all branches, and then either report all of them, or preferentially report errors from the patterns.  Or perhaps this is unnecessary and there is a way to backtrack with the elaboration of patterns that allows their errors to be reported early and also allows their typing to depend on the body of the branch.

Also, I believe the correct error for the second example is "the branch pattern b _ is redundant / overlaps with the catch-all idx," or something like that.

**Comment by JasonGross** [comment]
> Why does not Coq use the annotation that's on the definition? It seems to do that when the annotation is simpler (just nat instead of a match, like in the second example).

In the nat case, there is only one valid return clause: nat.  In the more complicated case, there are two valid return clauses: `match v as v1 return match v with ... end with` and `match v as v1 return match v1 with ... end with`.  Coq is not sure which you mean, and the unifier isn't strong enough to unify `nat` with `match ?v@{v1:=index} with
  | index | _ => nat
  end` even though there's a somewhat straightforward extension of small inversions that would let it do so.

**Comment by SkySkimmer** [comment]
If we try the "abstract the tycon" strategy first the output tests are improved (IMO) but for instance
~~~coq

Inductive listn : nat -> Set :=
  | niln : listn 0
  | consn : forall n : nat, nat -> listn n -> listn (S n).

Check
  fun (n : nat) (l : listn n) =>
   match l with
   | niln => l
   | consn n a c => l
   end.
~~~
starts failing. The problem seems to be a delayed constraint which causes the failure to be delayed until after we've decided that the strategy worked.
